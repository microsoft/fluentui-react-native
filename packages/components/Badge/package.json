{
  "name": "@fluentui-react-native/badge",
  "version": "0.4.21",
  "description": "A cross-platform Badge component using the Fluent Design System. A badge is an additional visual descriptor for UI elements.",
  "license": "MIT",
  "author": "Microsoft <fluentuinativeowners@microsoft.com>",
  "homepage": "https://github.com/microsoft/fluentui-react-native",
  "main": "src/index.ts",
  "module": "src/index.ts",
  "typings": "lib/index.d.ts",
  "onPublish": {
    "main": "lib-commonjs/index.js",
    "module": "lib/index.js"
  },
  "scripts": {
    "build": "fluentui-scripts build",
    "just": "fluentui-scripts",
    "clean": "fluentui-scripts clean",
    "lint": "fluentui-scripts eslint",
    "depcheck": "fluentui-scripts depcheck",
    "test": "fluentui-scripts jest",
    "update-snapshots": "fluentui-scripts jest -u",
    "prettier": "fluentui-scripts prettier",
    "prettier-fix": "fluentui-scripts prettier --fix true"
  },
  "dependencies": {
    "@fluentui-react-native/adapters": "^0.10.0",
    "@fluentui-react-native/experimental-shadow": ">=0.2.62 <1.0.0",
    "@fluentui-react-native/framework": ">=0.8.36 <1.0.0",
    "@fluentui-react-native/icon": "^0.17.7",
    "@fluentui-react-native/text": ">=0.19.15 <1.0.0",
    "@fluentui-react-native/theme-tokens": "^0.24.0",
    "@fluentui-react-native/theme-types": "^0.30.0",
    "@fluentui-react-native/tokens": "^0.20.5",
    "@fluentui-react-native/use-styling": "^0.9.1",
    "react-native-svg": "12.5.0"
  },
  "devDependencies": {
    "@fluentui-react-native/eslint-config-rules": "^0.1.1",
    "@fluentui-react-native/scripts": "^0.1.1",
    "@fluentui-react-native/test-tools": ">=0.1.1 <1.0.0",
    "@office-iss/react-native-win32": "^0.68.0",
    "@types/react-native": "^0.68.0",
    "react": "17.0.2",
    "react-native": "^0.68.0"
  },
  "peerDependencies": {
    "react": "17.0.2",
    "react-native": "^0.68.0"
  },
  "rnx-kit": {
    "reactNativeVersion": "^0.68",
    "reactNativeDevVersion": "^0.68",
    "kitType": "library",
    "capabilities": [
      "core",
      "core-android",
      "core-ios",
      "react",
      "svg"
    ]
  }
}
