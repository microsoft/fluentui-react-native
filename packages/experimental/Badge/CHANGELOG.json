{
  "name": "@fluentui-react-native/badge",
  "entries": [
    {
      "date": "Thu, 20 Jan 2022 21:05:25 GMT",
      "tag": "@fluentui-react-native/badge_v0.0.2",
      "version": "0.0.2",
      "comments": {
        "patch": [
          {
            "author": "v.kozlova13@gmail.com",
            "package": "@fluentui-react-native/badge",
            "commit": "ccea4990e137edde8d7c980ee498674b1a0a19bc",
            "comment": "Added Badge component."
          },
          {
            "author": "beachball",
            "package": "@fluentui-react-native/badge",
            "comment": "Bump @fluentui-react-native/experimental-text to v0.7.11",
            "commit": "bfb00a74fa8f5bcf5dc4d1a9da8cf8e76f0263b3"
          },
          {
            "author": "beachball",
            "package": "@fluentui-react-native/badge",
            "comment": "Bump @fluentui-react-native/icon to v0.11.5",
            "commit": "bfb00a74fa8f5bcf5dc4d1a9da8cf8e76f0263b3"
          },
          {
            "author": "beachball",
            "package": "@fluentui-react-native/badge",
            "comment": "Bump @fluentui-react-native/interactive-hooks to v0.11.13",
            "commit": "bfb00a74fa8f5bcf5dc4d1a9da8cf8e76f0263b3"
          },
          {
            "author": "beachball",
            "package": "@fluentui-react-native/badge",
            "comment": "Bump @fluentui-react-native/theme-tokens to v0.11.6",
            "commit": "bfb00a74fa8f5bcf5dc4d1a9da8cf8e76f0263b3"
          },
          {
            "author": "beachball",
            "package": "@fluentui-react-native/badge",
            "comment": "Bump @fluentui-react-native/tokens to v0.11.5",
            "commit": "bfb00a74fa8f5bcf5dc4d1a9da8cf8e76f0263b3"
          },
          {
            "author": "beachball",
            "package": "@fluentui-react-native/badge",
            "comment": "Bump @fluentui-react-native/use-styling to v0.8.2",
            "commit": "bfb00a74fa8f5bcf5dc4d1a9da8cf8e76f0263b3"
          }
        ]
      }
    }
  ]
}
