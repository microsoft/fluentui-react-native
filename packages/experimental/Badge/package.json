{
  "name": "@fluentui-react-native/badge",
  "version": "0.1.20",
  "description": "A cross-platform Badge component using the Fluent Design System. A badge is an additional visual descriptor for UI elements.",
  "license": "MIT",
  "author": "Microsoft <fluentuinativeowners@microsoft.com>",
  "homepage": "https://github.com/microsoft/fluentui-react-native",
  "main": "src/index.ts",
  "module": "src/index.ts",
  "typings": "lib/index.d.ts",
  "onPublish": {
    "main": "lib-commonjs/index.js",
    "module": "lib/index.js"
  },
  "scripts": {
    "build": "fluentui-scripts build",
    "just": "fluentui-scripts",
    "clean": "fluentui-scripts clean",
    "lint": "fluentui-scripts eslint",
    "depcheck": "fluentui-scripts depcheck",
    "test": "fluentui-scripts jest",
    "update-snapshots": "fluentui-scripts jest -u",
    "prettier": "fluentui-scripts prettier",
    "prettier-fix": "fluentui-scripts prettier --fix true"
  },
  "dependencies": {
    "@fluentui-react-native/adapters": "^0.9.0",
    "@fluentui-react-native/framework": ">=0.8.0 <1.0.0",
    "@fluentui-react-native/icon": "^0.13.4",
    "@fluentui-react-native/interactive-hooks": "^0.18.1",
    "@fluentui-react-native/text": ">=0.14.0 <1.0.0",
    "@fluentui-react-native/theme-tokens": "^0.19.0",
    "@fluentui-react-native/tokens": "^0.16.0",
    "@fluentui-react-native/use-styling": "^0.9.0",
    "react-native-svg": "^12.3.0"
  },
  "devDependencies": {
    "@fluentui-react-native/eslint-config-rules": "^0.1.1",
    "@fluentui-react-native/scripts": "^0.1.1",
    "@fluentui-react-native/test-tools": ">=0.1.1 <1.0.0",
    "@types/react-native": "^0.66.0",
    "react": "17.0.2",
    "react-native": "^0.66.0"
  },
  "peerDependencies": {
    "react": "17.0.2",
    "react-native": ">=0.66.0"
  }
}
