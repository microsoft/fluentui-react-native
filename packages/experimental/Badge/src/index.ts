export { badgeName } from './Badge.types';
export type {
  BadgeAppearance,
  BadgeProps,
  BadgeCoreProps,
  BadgeColor,
  BadgeShape,
  BadgeSize,
  BadgeSlotProps,
  BadgeCoreTokens,
  BadgeTokens,
  BadgeType,
  BadgeIconPosition,
} from './Badge.types';
export { BadgeAppearances, BadgeColors, BadgeShapes, BadgeSizes } from './Badge.types';
export { Badge, badgeLookup } from './Badge';
export { PresenceBadge, presenceBadgeName, PresenceBadgeStatuses } from './PresenceBadge';
export type { PresenceBadgeStatus, PresenceBadgeProps, PresenceBadgeSlotProps, PresenceBadgeType } from './PresenceBadge';
export { CounterBadge, counterBadgeName } from './CounterBadge';
export type { CounterBadgeProps, CounterBadgeSlotProps, CounterBadgeType } from './CounterBadge';
