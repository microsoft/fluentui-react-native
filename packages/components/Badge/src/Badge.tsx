/** @jsxRuntime classic */
/** @jsx withSlots */
import type { ReactNode } from 'react';
import { Children } from 'react';
import { View, I18nManager } from 'react-native';

import { Shadow } from '@fluentui-react-native/experimental-shadow';
import type { UseSlots } from '@fluentui-react-native/framework';
import { compose, withSlots, mergeProps } from '@fluentui-react-native/framework';
import { Icon, createIconProps } from '@fluentui-react-native/icon';
import { TextV1 as Text } from '@fluentui-react-native/text';

import { stylingSettings } from './Badge.styling';
import type { BadgeType, BadgeProps } from './Badge.types';
import { badgeName } from './Badge.types';
import { useBadge } from './useBadge';

export const badgeLookup = (layer: string, userProps: BadgeProps): boolean => {
  return (
    userProps[layer] ||
    layer === userProps['appearance'] ||
    layer === userProps['size'] ||
    (!userProps['size'] && layer === 'large') ||
    layer === userProps['shape'] ||
    (!userProps['shape'] && layer === 'circular') ||
    layer === userProps['badgeColor'] ||
    (I18nManager.isRTL && layer === 'rtl')
  );
};

export const Badge = compose<BadgeType>({
  displayName: badgeName,
  ...stylingSettings,
  slots: {
    root: View,
    icon: Icon,
    text: Text,
    shadow: Shadow,
  },
  useRender: (userProps: BadgeProps, useSlots: UseSlots<BadgeType>) => {
    const iconProps = createIconProps(userProps.icon);
    const badge = useBadge(userProps);

    const Slots = useSlots(userProps, (layer) => badgeLookup(layer, userProps));

    return (final: BadgeProps, ...children: ReactNode[]) => {
      const { icon, iconPosition, size, ...mergedProps } = mergeProps(badge, final);
      const showContent = size !== 'tiny' && size !== 'extraSmall';
      const showIcon = size !== 'tiny';
      return (
        <Slots.shadow>
          <Slots.root {...mergedProps}>
            {icon && showIcon && iconPosition === 'before' && <Slots.icon accessible={false} {...iconProps} />}
            {showContent &&
              Children.map(children, (child, i) =>
                typeof child === 'string' ? (
                  <Slots.text accessible={false} key={`text-${i}`}>
                    {child}
                  </Slots.text>
                ) : (
                  child
                ),
              )}
            {icon && showIcon && iconPosition === 'after' && <Slots.icon accessible={false} {...iconProps} />}
          </Slots.root>
        </Slots.shadow>
      );
    };
  },
});

export default Badge;
