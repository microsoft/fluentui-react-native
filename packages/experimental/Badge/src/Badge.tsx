/** @jsx withSlots */
import { Children, ReactNode } from 'react';
import { View, I18nManager } from 'react-native';
import { badgeName, BadgeType, BadgeProps } from './Badge.types';
import { TextV1 as Text } from '@fluentui-react-native/text';
import { compose, withSlots, UseSlots, mergeProps } from '@fluentui-react-native/framework';
import { Icon } from '@fluentui-react-native/icon';
import { createIconProps } from '@fluentui-react-native/interactive-hooks';
import { stylingSettings } from './Badge.styling';
import { useBadge } from './useBadge';

export const badgeLookup = (layer: string, userProps: BadgeProps): boolean => {
  return (
    userProps[layer] ||
    layer === userProps['appearance'] ||
    (!userProps['appearance'] && layer === 'filled') ||
    layer === userProps['size'] ||
    (!userProps['size'] && layer === 'large') ||
    layer === userProps['shape'] ||
    (!userProps['shape'] && layer === 'circular') ||
    layer === userProps['badgeColor'] ||
    (!userProps['badgeColor'] && layer === 'brand') ||
    (I18nManager.isRTL && layer === 'rtl')
  );
};

export const Badge = compose<BadgeType>({
  displayName: badgeName,
  ...stylingSettings,
  slots: {
    root: View,
    icon: Icon,
    text: Text,
  },
  useRender: (userProps: BadgeProps, useSlots: UseSlots<BadgeType>) => {
    const iconProps = createIconProps(userProps.icon);
    const badge = useBadge(userProps);

    const Slots = useSlots(userProps, (layer) => badgeLookup(layer, userProps));

    return (final: BadgeProps, ...children: ReactNode[]) => {
      const { icon, iconPosition = 'before', ...mergedProps } = mergeProps(badge, final);
      return (
        <Slots.root {...mergedProps}>
          {icon && iconPosition === 'before' && <Slots.icon accessible={false} {...iconProps} />}
          {Children.map(children, (child, i) =>
            typeof child === 'string' ? (
              <Slots.text accessible={false} key={`text-${i}`}>
                {child}
              </Slots.text>
            ) : (
              child
            ),
          )}
          {icon && iconPosition === 'after' && <Slots.icon accessible={false} {...iconProps} />}
        </Slots.root>
      );
    };
  },
});

export default Badge;
