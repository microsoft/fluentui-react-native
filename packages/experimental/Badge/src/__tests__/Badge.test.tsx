import * as React from 'react';
import { Badge } from '../Badge';
import * as renderer from 'react-test-renderer';
import { checkRenderConsistency, checkReRender } from '@fluentui-react-native/test-tools';

describe('Badge component tests', () => {
  const fontBuiltInProps = {
    fontFamily: 'Arial',
    codepoint: 0x2663,
    fontSize: 16,
  };
  it('Empty Badge', () => {
    const tree = renderer.create(<Badge />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Badge all props', () => {
    const tree = renderer
      .create(
        <Badge
          text="Badge"
          size="large"
          appearance="outline"
          shape="rounded"
          icon={{ fontSource: { ...fontBuiltInProps }, color: '#fff' }}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Badge tokens', () => {
    const BadgeStyled = Badge.customize({
      backgroundColor: 'yellow',
      borderColor: '#f09',
      borderWidth: 4,
    });
    const tree = renderer.create(<BadgeStyled text="Badge Tokens" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Check checkRenderConsistency for Badge', () => {
    checkRenderConsistency(() => <Badge />, 2);
  });

  it('Badge re-renders correctly', () => {
    checkReRender(() => <Badge />, 2);
  });
});
