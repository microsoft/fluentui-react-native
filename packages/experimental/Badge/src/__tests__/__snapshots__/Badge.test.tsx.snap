// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge component tests Badge all props 1`] = `
<View
  appearance="outline"
  shape="rounded"
  size="large"
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "backgroundColor": "transparent",
      "borderColor": "#0078d4",
      "borderRadius": 4,
      "borderWidth": 1,
      "bottom": 0,
      "flexDirection": "row",
      "justifyContent": "center",
      "minHeight": 24,
      "minWidth": 24,
      "paddingHorizontal": 4,
      "right": 0,
      "width": undefined,
    }
  }
>
  <Text
    accessible={false}
    style={
      Object {
        "color": "#fff",
        "fontFamily": "Arial",
        "fontSize": 16,
        "fontWeight": "400",
        "margin": 0,
      }
    }
  >
    ♣
  </Text>
  <Text
    accessible={false}
    ellipsizeMode="tail"
    fontFamily="Segoe UI"
    fontSize={12}
    fontWeight="600"
    numberOfLines={0}
    paddingHorizontal={2}
    style={
      Object {
        "color": "#0078d4",
        "fontFamily": "Segoe UI",
        "fontSize": 12,
        "fontStyle": undefined,
        "fontWeight": "400",
        "margin": 0,
        "textAlign": undefined,
        "textDecorationLine": undefined,
      }
    }
  >
    Badge
  </Text>
</View>
`;

exports[`Badge component tests Badge tokens 1`] = `
<View
  size="medium"
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "backgroundColor": "#0078d4",
      "borderColor": "transparent",
      "borderRadius": 4,
      "borderWidth": 4,
      "bottom": 0,
      "flexDirection": "row",
      "justifyContent": "center",
      "minHeight": 24,
      "minWidth": 24,
      "paddingHorizontal": 4,
      "right": 0,
      "width": undefined,
    }
  }
>
  <Text
    accessible={false}
    ellipsizeMode="tail"
    fontFamily="Segoe UI"
    fontSize={12}
    fontWeight="600"
    numberOfLines={0}
    paddingHorizontal={2}
    style={
      Object {
        "color": "#ffffff",
        "fontFamily": "Segoe UI",
        "fontSize": 12,
        "fontStyle": undefined,
        "fontWeight": "400",
        "margin": 0,
        "textAlign": undefined,
        "textDecorationLine": undefined,
      }
    }
  >
    Badge Tokens
  </Text>
</View>
`;

exports[`Badge component tests Empty Badge 1`] = `
<View
  size="medium"
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "backgroundColor": "#0078d4",
      "borderColor": "transparent",
      "borderRadius": 4,
      "borderWidth": 1,
      "bottom": 0,
      "flexDirection": "row",
      "justifyContent": "center",
      "minHeight": 24,
      "minWidth": 24,
      "paddingHorizontal": 4,
      "right": 0,
      "width": undefined,
    }
  }
/>
`;

exports[`CounterBadge component tests CounterBadge all props 1`] = `
<View
  appearance="ghost"
  shape="rounded"
  size="large"
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "backgroundColor": undefined,
      "borderRadius": 4,
      "borderWidth": 1,
      "bottom": 0,
      "flexDirection": "row",
      "justifyContent": "center",
      "minHeight": 24,
      "paddingHorizontal": 4,
      "right": 0,
      "width": undefined,
    }
  }
>
  <Text
    accessible={true}
    style={
      Object {
        "color": "#fff",
        "fontFamily": "Arial",
        "fontSize": 16,
        "fontWeight": "400",
        "margin": 0,
      }
    }
  >
    ♣
  </Text>
  <Text
    ellipsizeMode="tail"
    numberOfLines={0}
    style={
      Object {
        "color": "#323130",
        "fontFamily": "Segoe UI",
        "fontSize": 12,
        "fontStyle": undefined,
        "fontWeight": "400",
        "margin": 0,
        "textAlign": undefined,
        "textDecorationLine": undefined,
      }
    }
  >
    30
  </Text>
</View>
`;

exports[`CounterBadge component tests CounterBadge tokens 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "backgroundColor": "#0078d4",
      "borderColor": "transparent",
      "borderRadius": 4,
      "borderWidth": 4,
      "bottom": 0,
      "flexDirection": "row",
      "justifyContent": "center",
      "minHeight": 24,
      "paddingHorizontal": 4,
      "right": 0,
      "width": undefined,
    }
  }
>
  <Text
    ellipsizeMode="tail"
    numberOfLines={0}
    style={
      Object {
        "color": "#ffffff",
        "fontFamily": "Segoe UI",
        "fontSize": 12,
        "fontStyle": undefined,
        "fontWeight": "400",
        "margin": 0,
        "textAlign": undefined,
        "textDecorationLine": undefined,
      }
    }
  >
    undefined
  </Text>
</View>
`;

exports[`CounterBadge component tests Empty Badge 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "backgroundColor": "#0078d4",
      "borderColor": "transparent",
      "borderRadius": 4,
      "borderWidth": 1,
      "bottom": 0,
      "flexDirection": "row",
      "justifyContent": "center",
      "minHeight": 24,
      "paddingHorizontal": 4,
      "right": 0,
      "width": undefined,
    }
  }
>
  <Text
    ellipsizeMode="tail"
    numberOfLines={0}
    style={
      Object {
        "color": "#ffffff",
        "fontFamily": "Segoe UI",
        "fontSize": 12,
        "fontStyle": undefined,
        "fontWeight": "400",
        "margin": 0,
        "textAlign": undefined,
        "textDecorationLine": undefined,
      }
    }
  >
    undefined
  </Text>
</View>
`;

exports[`PresenceBadge component tests PresenceBadge props 1`] = `
<View
  iconPosition="before"
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "aspectRatio": 1,
      "backgroundColor": "#ffffff",
      "borderColor": "#ffffff",
      "borderWidth": 2,
      "bottom": -2,
      "flexDirection": "row",
      "height": 24,
      "justifyContent": "center",
      "paddingHorizontal": 4,
      "position": "absolute",
      "right": -2,
      "width": 24,
    }
  }
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={20}
    bbWidth={20}
    color={-15490802}
    fill="none"
    focusable={false}
    meetOrSlice={0}
    minX={0}
    minY={0}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "aspectRatio": 1,
          "color": "#13a10e",
          "height": 20,
          "width": 20,
        },
        Object {
          "flex": 0,
          "height": 20,
          "width": 20,
        },
      ]
    }
    tintColor={-15490802}
    vbHeight={16}
    vbWidth={16}
  >
    <RNSVGGroup
      fill={null}
      propList={
        Array [
          "fill",
        ]
      }
    >
      <RNSVGPath
        d="M11.7 6.7a1 1 0 00-1.4-1.4L7 8.58l-1.3-1.3a1 1 0 00-1.4 1.42l2 2a1 1 0 001.4 0l4-4zM0 8a8 8 0 1116 0A8 8 0 010 8zm8-6a6 6 0 100 12A6 6 0 008 2z"
        fill={
          Array [
            2,
          ]
        }
        propList={
          Array [
            "fill",
          ]
        }
      />
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;
